### ISSUE TITLE
datastore: PropertyLoadSaver with context

### ISSUE THREAD

**Issue by benjaminjkraft** [googleapis/google-cloud-go#1718]
We find ourselves wanting to use the `context.Context` in our `Load`/`Save` methods.  Sadly, this doesn't seem to be possible currently.  Would it be possible to add a way to do this?

The way I would suppose to implement this is something like the following.  The datastore package would define an additional interface
```go
type ContextPropertyLoadSaver interface {
	Load(context.Context, []Property) error
	Save(context.Context) ([]Property, error)
}
```
The functions that currently use PropertyLoadSaver would then check both (underneath they might convert one into the other).  (I'm pretty sure all of those functions already get passed a context.)  One could also add a similar version of KeyLoader.

**Comment by benjaminjkraft** [comment]
A bit more context as to why this would be helpful: we would like our tests to run using some sort of fake-time, so that they can run reproducibly and make nontrivial assertions relating to time.  The details of how we will do so are not particularly relevant, but ideally we'd avoid using a global (which for example makes it harder to run tests in parallel).  However, perhaps the most common place we check the current time is in `PropertyLoadSaver` methods, wherein we set creation/modification timestamp fields to the current time.  (Nearly all of our models have at least a modification-timestamp field.)

So we need some way to get a source-of-time to our `PropertyLoadSaver` methods; and sadly we don't see any way other than changing the datastore library.  It seems like the most natural generalizable way to do so would be for it to accept context; then we can use `context.WithValue` to put whatever we need in there.  (Ideally, we might have it simply accept arbitrary arguments which could somehow be passed to relevant datastore API calls, or something like that, but that seems like a nightmare of an API even if one had generics with which to write it!)

**Comment by tbpg** [comment]
Would it be possible to do something like this and avoid needing to extend the `datastore` package? The general idea is to store the source of time in your `PropertyLoadSaver` and to supply that source when you initialize the `PropertyLoadSaver` for tests. If the source is `nil`, use `time.Now()`.

This doesn't appear to be more complicated than using a `context.Context` to pass the source of time, even if there was a built-in `ContextPropertyLoadSaver` interface.

```go
type myPropLoadSaver struct {
	now func() time.Time
}

func (m *myPropLoadSaver) Load([]Property) error {
	now := time.Now()
	if m.now != nil {
		now = m.now()
	}
	_ = now
	// ...
	return nil
}

func (m *myPropLoadSaver) Save() ([]Property, error) {
	now := time.Now()
	if m.now != nil {
		now = m.now()
	}
	_ = now
	// ...
	return nil, nil
}

func testingPropertyLoadSaver(now func() time.Time) PropertyLoadSaver {
	return &myPropLoadSaver{now: now}
}
```

Of course, you could make a separate type/interface for `func() time.Time`, but when writing this up I found that increased verbosity without adding clarity.

**Comment by benjaminjkraft** [comment]
That's an interesting idea, and not one we had considered!  I think that, or something similar, will resolve things to our satisfaction.  It's not quite as clean as we'd like, but I understand the hesitation to add another similar interface just for an uncommon use case.  So feel free to close if the below is not of interest (or I can add it to a separate new issue if you prefer).

One thing we were thinking about along the way, which may be of interest, is that really, what we want to do doesn't need the full power of `PropertyLoadSaver` -- it doesn't want to change anything about how properties get serialized, just modify their values before save.  So we could have an interface
```go
type PreSaver interface {
	PreSave(ctx context.Context) error
}
```
(and likewise for `PreLoader`) and the idea is the model can implement that method if it wants to get a chance to clean things up before save (it would just modify the properties in place).

This is a bit simpler to think about for such cases -- for example it plays a lot nicer with embedding: if you want to embed some other type that has a `PreSaver`, and you don't want to do any pre-save logic of your own, you can just embed that type; `PropertyLoadSaver` you have to wrap its methods just to make them do the default for your properties.  (This is in fact the case in our codebase -- we have a shared base-model that includes some standard metadata like modification date for backup-management purposes.)  I suspect it could handle many of the common usages of `PropertyLoadSaver` -- certainly both the ones in our codebase and the one in the documentation.  Of course, the two should play fine together if you do want both behaviors.

In principle this has nothing to do with the original issue!  But it's a convenient place to add `context` without having to have two parallel interfaces whose only difference is `context`.

We could implement this in user-land, something like the below.  But obviously it's easier to make sure it's right and kept up to date with client changes if it's upstreamed.
<details>
<summary>User-land implementation</summary>

```go
func (c WrappedClient) callPreSave(ctx context.Context, model interface{}) error {
	// model might be a single model -- in which case this is easy
	ps, ok := model.(PreSaver)
	if ok {
		return ps.PreSave(ctx)
	}

	// or a slice -- in which case we have to use reflect
	rv := reflect.ValueOf(model)
	if rv.Kind() == reflect.Slice {
		// Equivalent to `for i, item := range model`
		for i := 0; i < rv.Len(); i++ {
			item := rv.Index(i).Interface()
			ps, ok = item.(PreSaver)
			if ok {
				err := ps.PreSave(ctx)
				if err != nil {
					return err
				}
			}
		}
	}

	return nil
}

func (c WrappedClient) Put(ctx context.Context, key *Key, src interface{}) (*Key, error) {
	c.callPreSave(ctx, src)
	return c.Client.Put(ctx, key, src)
}

// similarly for PutMulti, and probably NewInsert/NewUpdate/NewUpsert; and similarly for callPreLoad on Get/GetMulti/etc.
```
</details>

**Comment by tbpg** [comment]
I'm glad that (or a variant) works for you. Let us know if you end up needing this for something else.

Feel free to file another FR for the `PreSaver` idea. It does seem like it would be useful, but as you said it's possible to implement in user-land. cc @BenWhitehead @tritone, who work on Datastore/Firestore (among other things). If we did this for Go, we might want something similar across all languages. So, I'm curious if there have been any other requests for something like this.

I'm going to close this issue for now. If you end up implementing that in user-land and it is substantially nicer to use than just `PropertyLoadSaver`, let us know!

**Comment by benjaminjkraft** [comment]
I finally got around to trying to implement this, and the suggestion doesn't work as well as we had hoped.  In particular, the issue is: suppose you have code like this:
```go
type MyModel struct {
  ModifiedDate time.Time `datastore:"created_date"`
  // ...
}

func UpdateMany(keys []*datastore.Key) error {
  ms := make([]*MyModel, len(keys))
  err := ctx.Datastore().Get(ctx, key, ms)
  // check error, modify ms
  _, err = ctx.Datastore().Put(ctx, key, ms)
  return err
}
```

We want to use the propertyloadsaver trick here; and we want to test this update function.  So we add a `now` field, and `Load` and `Save` methods (or embed some struct which does so) and change the signature to
```go
func Update(now func() time.Time, keys []*datastore.Key) error
```
(or pass in `ctx`, or whatever).  The problem is: what do we do with `now`?  When we were just putting a single model, it was pretty natural: we replaced `m := MyModel{<properties>}` with `m := MyModel{now: now, <properties>}`.  Now we have to replace the simple `make` with a loop to set `now` on each model, which reads very strangely (why are we doing all this work to set up the result-value for `Get`?), or we have to set it in between `Get` and `Put` (where we are probably already looping), which a less natural place to audit (in code review or programmatically): maybe now we just forget to set it.  (This is especially important for us where we have a base-model with a timestamp (used for incremental exports), so the application code doesn't even necessarily want to care about this.

So I think we're looking back at wrapping the client.  We're not the most excited about making sure we get all the methods right, so I wanted to check back first to see if there are any more ideas for how to solve this in simpler ways!

**Comment by tbpg** [comment]
Could `Update` take `[]*MyModel` as an argument? I'm not sure what the calling code looks like, but I'm curious where `[]*datastore.Key` comes from.

If the odd `for` loop was in its own function, perhaps it wouldn't be so bad? `setSourceOfTime(sourceOfTime func() time.Time, ms *MyModel...)`. If you want to test `ModifiedDate`, call that function?

`PreSave` does seem like it would solve this problem. You attach the source of time to the `context.Context`, pass that to `Get`/`Put`, then `Put` calls `PreSave`, which pulls out the source of time.

I'm also not sure about storing the Datastore client inside the context. It's not a request scoped value, so I'm curious about that. (Not saying it's wrong, just caught my eye.) Is there a `context.Context` arg to `Update`?

**Comment by benjaminjkraft** [comment]
I'm generalizing from several examples, but a common pattern in our code might be that those keys come from some other model (we have a classroom model, and it stores a list of students) or the `Get` is actually a `Query` and we want to update all matching entities.  In the former case at least, `Update` could take the `[]*MyModel` as an argument, but eventually somebody has to fetch it, and that somebody plausibly might want to be tested.  I guess the further you push it out, the less likely it matters that the now-function is not settable.

I do like the idea that you provide a set-time-on-many function.  But it doesn't really solve the problem that there's no good point to check that you didn't forget to do the updating.  I guess, you go to write a test and you realize you can't?  It just feels like trouble to us.

> I'm also not sure about storing the Datastore client inside the context.

Whoops, this is not actually relevant to the example, I just forgot to switch my example back to the standard style, where you would do `client.Get()` and such, and client would be an argument or global or whatever.  (Putting it in context is something we're trying, essentially by way of dependency injection.  It's working out pretty well so far, although it's taken us some iteration to make the right choices, and I definitely don't think it makes sense for everyone.  We may write more about it in a blog post eventually.)  Anyway, there would be a ctx arg to Update with or without that, I just forgot to include it!

**Comment by tbpg** [comment]
I'm asking around to see if anyone else has any other ideas. If this is urgent, let us know.

**Comment by benjaminjkraft** [comment]
Much appreciated.  It's not urgent -- the sooner we have a good solution the easier it is for us, but also if it gets too frustrating we'll implement `PreSaver` ourselves.  (It's messier than I thought w/r/t mutations and transactions, which we would have to wrap, which is why I haven't already done so.)

**Comment by benjaminjkraft** [comment]
Thanks!  Not particularly urgent -- or rather if it becomes bothersome enough for us we'll do `PreSaver` in user-land.  (It's messier than I thought, because we have to wrap the transaction and mutation objects, but still doable I think.)

**Comment by tbpg** [comment]
I reached out internally for other ideas, but the consensus is that storing the source of time as an explicit field in your struct is the way to go. It may be tedious to set and update, but it seems better than plumbing arbitrary values through a `context.Context`.

One idea was to have a constructor that sets the source of time. That way, you don't have to do a check every time the `Load`/`Save` methods are called.

I'm going to re-close this. But, let us know if something else comes up.
